File: 0001-two-sum.py
"""
1. Two Sum

Submitted: January 16, 2025

Runtime: 0 ms (beats 100.00%)
Memory: 19.00 MB (beats 19.63%)
"""

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = {}
        return next(x for x in (
            (i, d[target - nums[i]]) if ((target - nums[i]) in d) else d.update({ nums[i]: i })
            for i in range(len(nums))
            ) if x is not None
        )

File: 0001-two-sum.cpp
/*
1. Two Sum

Submitted: December 20, 2024

Runtime: 3 ms (beats 77.60%)
Memory: 14.80MB (beats 24.95%)
*/

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> map;
        for (int i = 0, n = nums.size(); i < n; ++i) {
            if (map.count(target - nums[i])) {
                return { i, map[target - nums[i]] };
            } else {
                map[nums[i]] = i;
            }
        }
        return {}; // this is only here so that compliation doesn't fail
    }
};